### PR TITLE
Feature/retrieval download modals

### PR DESCRIPTION
**Pull request by thiessenp-cds**
# Summary | Résumé

This contains the work for the Responses page and the related sub features. This feature combines work from most of the team also so this is true teamwork :)

| Before                                          | After                                        |
| ----------------------------------------------- | -------------------------------------------- |
| N/A |  Screenshot 2023-03-30 at 2 42 10 PM |
| N/A | Screenshot 2023-03-30 at 3 46 55 PM  |
| N/A | Screenshot 2023-03-30 at 3 47 06 PM |
| N/A (top) |  Screenshot 2023-03-31 at 10 02 58 AM |
| N/A (bottom) | Screenshot 2023-03-31 at 10 02 51 AM  |
| N/A (top) | Screenshot 2023-03-31 at 10 03 37 AM  |
| N/A (bottom) |  Screenshot 2023-03-31 at 10 03 44 AM |
| N/A |  Screenshot 2023-03-31 at 9 42 51 AM |
| N/A | Screenshot 2023-03-31 at 9 43 11 AM  |
| N/A |  Screenshot 2023-03-31 at 9 55 41 AM |
| N/A | Screenshot 2023-03-31 at 9 59 43 AM |
| N/A | Screenshot 2023-03-31 at 10 01 08 AM |

# Test instructions | Instructions pour tester la modification

There is a lot to test. Try creating a form with at least 21 submissions and try to find edge cases that break the Download Table.

# Unresolved questions / Out of scope | Questions non résolues ou hors sujet

The designs didn't include some error cases and I added them along with the language strings.

# Pull Request Checklist

Please complete the following items in the checklist before you request a review:

- [x] Have you completely tested the functionality of change introduced in this PR? Is the PR solving the problem it's meant to solve within the scope of the related issue?
- [x] The PR does not introduce any new issues such as failed tests, console warnings or new bugs.
- [ ] If this PR adds a package have you ensured its licensed correctly and does not add additional security issues?
- [x] Is the code clean, readable and maintainable? Is it easy to understand and comprehend.
- [x] Does your code have adequate comprehensible comments? Do new functions have [docstrings](https://tsdoc.org/)?
- [ ] Have you modified the change log and updated any relevant documentation?
- [x] Is there adequate test coverage? Both unit tests and end-to-end tests where applicable?
- [x] If your PR is touching any UI is it accessible? Have you tested it with a screen reader? Have you tested it with automated testing tools such as axe?

### CODE REVIEW

File: public/static/locales/fr/form-builder-responses.json [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      "downloading":"Téléchargement",
+      "downloadComplete": "Téléchargement complet",
+      "downloadingXFiles": "Téléchargement de {{fileCount}} fichiers",
+      "downloadXSelectedResponses": "Télécharger {{size}} réponses sélectionnées",
+      "trySelectingLessFilesHeader": "Seules {{max}} réponses peuvent être téléchargées à la fois.",
+      "trySelectingLessFiles":"Vous avez {{max}} réponses sélectionnées. Décochez au moins 1 réponse avant de télécharger.",
+      "atLeastOneFileHeader": "Au moins un fichier doit être sélectionné à télécharger.",
+      "atLeastOneFile": "Essayez de sélectionner au moins un fichier à télécharger.",
+      "errorDownloadingFilesHeader":"Erreur lors du téléchargement des fichiers sélectionnés.",
+      "errorDownloadingFiles":"Réessayez dans quelques minutes. Si le problème persiste, contactez ",
```
Comment (thiessenp-cds) [inline public/static/locales/fr/form-builder-responses.json:55]: Inline links could also be accomplished with the Trans component but I wanted to hold off adding too many new dependencies/ways of doing things for one feature: https://react.i18next.com/latest/trans-component

File: public/static/locales/fr/form-builder-responses.json [hunk 2]
```diff
@@ -0,0 +1,97 @@
+{
```
Comment (thiessenp-cds) [inline public/static/locales/fr/form-builder-responses.json:1]: Moved out of form-builder.json because of the many conflicts I kept getting

File: lib/clientHelpers.ts [hunk 3]
```diff
@@ -13,6 +13,10 @@ import {
 import { Submission } from "@lib/types/submission-types";
```
Comment (thiessenp-cds) [inline lib/clientHelpers.ts:13]: Renamed helpers.ts to clientHelpers.ts to help make it clear that client util functions can go here. The rest of the lib directory is for server related util functions

File: lib/clientHelpers.ts [hunk 4]
```diff
@@ -244,6 +248,85 @@ function _rehydrateCheckBoxResponse(response: Response) {
...
+  // Mainly for unit testing - default to today's date if no date is passed (main case)
+  let date1 = null;
+  if (!startDate) {
+    date1 = new Date();
+  } else if (startDate instanceof Date) {
+    date1 = startDate;
+  } else if (typeof startDate === "number" && String(startDate).length === 13) {
+    date1 = new Date(startDate);
+  } else {
+    return -1; // Invalid date
```
Comment (thiessenp-cds) [inline lib/clientHelpers.ts:314]: Using as an error flag so it's easier to handle in template code vs throwing an error

File: components/form-builder/app/shared/Toast.tsx [hunk 5]
```diff
@@ -85,7 +85,7 @@ export const toast = {
     originalToast.error(<p className="py-2">{message}</p>);
   },
   info: (message: string) => {
-    originalToast.info(<p className="py-2">{message}</p>);
+    originalToast.info(<p className="py-1">{message}</p>);
```
Comment (bryan-robitaille) [inline components/form-builder/app/shared/Toast.tsx:88]: Assuming it's just an oversight but why change the padding for only the Info styling?

File: components/form-builder/app/shared/Toast.tsx [hunk 6]
```diff
@@ -96,4 +96,7 @@ export const toast = {
   default: (message: string) => {
     originalToast(<p className="py-2">{message}</p>);
   },
+  dismiss: (id?: string | undefined) => {
```
Comment (bryan-robitaille) [inline components/form-builder/app/shared/Toast.tsx:99]: Where is this new function called in the code?  I can seem to find any reference to it.

**Review by bryan-robitaille — CHANGES_REQUESTED** [review]
Two super quick things I noticed while skimming through the code.

File: components/form-builder/app/responses/DownloadTable.tsx [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+
+// TODO: move to an app setting variable
+const MAX_FILE_DOWNLOADS = 20;
+
+// Using a reducer to have more control over when the template is updated (reduces re-renders)
+const reducerTableItems = (state: ReducerTableItemsState, action: ReducerTableItemsActions) => {
+  switch (action.type) {
+    case "UPDATE": {
+      if (!action.item) {
+        throw Error("Table dispatch missing item checkox state");
```
Comment (timarney) [inline components/form-builder/app/responses/DownloadTable.tsx:54]:
```suggestion
        throw Error("Table dispatch missing item checkbox state");
```

**Comment by timarney** [comment]
Some thoughts for follow-up PRs

1) When an error occurs for report a problem the dialog should scroll up to the error message

Screen Shot 2023-04-04 at 8 14 55 AM

2) 'cancel' is all lower case vs `Report problem`

3) Confirm receipt - cancel is all lower case

4) Confirm receipt. I entered `c0c42c90-b816-49da-a69a-80bfd8ba15db`  - press confirm  receipt --- first press moves to entered a second code confirm  receipt again closes the dialog without telling me anything happened.

File: components/form-builder/app/responses/DownloadResponseStatus.tsx [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+}: {
+  vaultStatus: string;
+  createdAt: Date | number;
+  downloadedAt?: Date | number;
+  overdueAfter?: number;
+}) => {
+  const { t } = useTranslation("form-builder-responses");
+  let status = null;
+
+  if (vaultStatus === "New") {
```
Comment (timarney) [inline components/form-builder/app/responses/DownloadResponseStatus.tsx:21]:
```suggestion
  if (vaultStatus === VaultStatus.NEW) {
```

File: lib/clientHelpers.ts [hunk 9]
```diff
@@ -244,6 +248,85 @@ function _rehydrateCheckBoxResponse(response: Response) {
...
+export const randomId = () => {
+  return Math.random().toString(36).substr(2, 9);
+};
+
+/**
+ * Format date for: YYYY-MM-DD
+ * @param date to format. Expects a Date type
+ * @returns date formatted
+ */
+export const formatDate = (date: Date): string => {
```
Comment (timarney) [inline lib/clientHelpers.ts:286]: Maybe use:

```javascript
const date = new Date();
date.toLocaleDateString("en-CA", {
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  });
 ```

File: lib/clientHelpers.ts [hunk 10]
```diff
@@ -244,6 +248,85 @@ function _rehydrateCheckBoxResponse(response: Response) {
...
+  const date2 =
+    typeof endDate === "number" && String(endDate).length === 13 ? new Date(endDate) : endDate;
+
+  if (!(date2 instanceof Date)) {
+    return -1; // Invalid date
+  }
+
+  const daysDiff = Math.abs(Number(date2) - Number(date1));
+  const daysPassed = Math.round(daysDiff / (1000 * 60 * 60 * 24));
+  return daysPassed;
```
Comment (timarney) [inline lib/clientHelpers.ts:327]: Didn't confirm but I think you might end up with NaN here if date1 ends up keeping it `null` value (somehow)

File: lib/nagware.ts [hunk 11]
```diff
@@ -1,62 +1,83 @@
...
 
-export function detectOldUnprocessedSubmissions(submissions: NagwareSubmission[]): NagwareResult {
-  const currentDate = Date.now();
+export async function detectOldUnprocessedSubmissions(
+  submissions: NagwareSubmission[]
+): Promise<NagwareResult> {
+  try {
+    const currentDate = Date.now();
+    const [promptPhaseDays, warnPhaseDays] = await Promise.all([
+      getAppSetting("nagwarePhasePrompted"),
```
Comment (timarney) [inline lib/nagware.ts:11]: Should probably default to something if this fails?

Comment (bryan-robitaille) [reply lib/nagware.ts:11]: There is a check just below so that if either promptPhaseDays or warnPhaseDays are null/undefined the function immediately returns a `NagLevel.None` so the client will not show any messages.  It's like if the feature is turned off.

File: components/form-builder/app/shared/Toast.tsx (hunk 5 above)
Comment (thiessenp-cds) [reply components/form-builder/app/shared/Toast.tsx:88]: I think that's right, just an oversight. At one point I was trying to "squish" the table cells as close as possible together - it's probably left over from that

File: components/form-builder/app/shared/Toast.tsx (hunk 6 above)
Comment (thiessenp-cds) [reply components/form-builder/app/shared/Toast.tsx:99]: Yes, if I remember right, the implementation I had, had a manual dismiss. Dave added it for me I believe. Then the dismiss functionality was removed from my code. So it can be removed - heh it's almost like an awkward. conversation that was left hanging :)

**Comment by thiessenp-cds** [comment]
> Some thoughts for follow-up PRs
> 
> 1. When an error occurs for report a problem the dialog should scroll up to the error message
> 
> Screen Shot 2023-04-04 at 8 14 55 AM
> 
> 2. 'cancel' is all lower case vs `Report problem`
> 3. Confirm receipt - cancel is all lower case
> 4. Confirm receipt. I entered `c0c42c90-b816-49da-a69a-80bfd8ba15db`  - press confirm  receipt --- first press moves to entered a second code confirm  receipt again closes the dialog without telling me anything happened.
> 
> Checking the request I see an error happened
> 
> ```
> {"invalidConfirmationCodes":["c0c42c90-b816-49da-a69a-80bfd8ba15db"]}
> ```
> 
> 5. Input for more entries likely shouldn't show or at least now allow for inputting more when at the limit.
> 
> Screen Shot 2023-04-04 at 8 35 15 AM

Makes sense. I think more behaviour needs to be designed, especially error messages. But I agree, no error message about an invalid code is very confusing. I'll add it and we can iterate/remove it later if we want to

**Review by timarney — APPROVED** [review]
Approving --- based on initial testing noted a few things here:
These can be in follow up PRs 
https://github.com/cds-snc/platform-forms-client/pull/1795#issuecomment-1495879354

**Review by timarney — APPROVED** [review]

**Review by bryan-robitaille — APPROVED** [review]
